File: backend/golang/models/GpMessage.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// GpMessage struct for MongoDB document
type GpMessage struct {
	ID        primitive.ObjectID `bson:"_id,omitempty" json:"_id,omitempty"` // Optional
	UUID      string             `bson:"uuId" json:"uuId" validate:"required"`
	IsFile    bool               `bson:"isFile" json:"isFile" validate:"required"`
	From      string             `bson:"from" json:"from" validate:"required"`
	To        string             `bson:"to" json:"to" validate:"required"`
	Text      string             `bson:"text" json:"text" validate:"required"`
	CreatedAt time.Time          `bson:"createdAt" json:"createdAt" validate:"required"`
	UpdatedAt time.Time          `bson:"updatedAt" json:"updatedAt" validate:"required"`
}

File: backend/nodejs/src/db/mongodb/models/GpMessage.ts
import mongoose, { Schema, Document, Model } from "mongoose";
import { IUser } from "@/db/mongodb/models/User";
import { IGroup } from "@/db/mongodb/models/Group";

// Define an interface representing a Group Message document
export interface IGpMessage {
  _id: mongoose.Types.ObjectId;
  uuId: string;
  isFile: boolean;
  from: string | IUser;
  to: string | IGroup;
  text: string;
  createdAt: Date;
  updatedAt: Date;
}

export type KafkaIGpMessageType = Omit<IGpMessage, keyof Document | "_id" | "isFile" | "from" | "to" | "updatedAt"> & {
  from: string;
  to: string;
  isFile?: boolean;
};

// Define the Group Message schema using the interface
const gpMessageSchema = new Schema<IGpMessage>({
  uuId: {
    type: String,
    required: true,
    unique: true,
  },
  isFile: {
    type: Boolean,
    default: false,
  },
  from: {
    type: String,
    ref: "User",
    required: true,
  },
  to: {
    type: String,
    ref: "Group",
    required: true,
  },
  text: {
    type: String,
    required: true,
    trim: true,
  },
  createdAt: {
    type: Date,
    required: true,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
});

// Create the Group Message model
const GpMessage: Model<IGpMessage> = mongoose.model<IGpMessage>("GpMessage", gpMessageSchema);

export default GpMessage;
